File: binary-tree/reverseLevelOrderTraversal.js
/*
  Given a binary tree, populate an array to represent its level-by-level traversal in reverse order, i.e., the lowest level comes first. You should populate the values of all nodes in each level from left to right in separate sub-arrays.
*/

class TreeNode {

  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
};

const reverseLevelOrderTraversal = function (root) {
  const result = [];
  const queue = [root];
  if (!root) return result;
  while (queue.length > 0) {
    const currentLevel = [];
    let length = queue.length;
    let count = 0;
    while (count < length) {
      const currentNode = queue.shift();
      if (currentNode.left) queue.push(currentNode.left);
      if (currentNode.right) queue.push(currentNode.right);
      currentLevel.push(currentNode.value);
      count++
    }
    result.unshift(currentLevel);
  }
  return result;
}

var root = new TreeNode(12)
root.left = new TreeNode(7)
root.right = new TreeNode(1)
root.left.left = new TreeNode(9)
root.right.left = new TreeNode(10)
root.right.right = new TreeNode(5)
console.log(`Reverse level order traversal: ${reverseLevelOrderTraversal(root)}`)

File: java/tree_bfs/ReverseLevelOrderTraversal.java
/*
  @requirement: Given a binary tree, find a reverse order traversal of the given binary tree
  @param: TreeNode: tree
  @return: ArrayList<ArrayList<Integer>>
*/

import java.util.LinkedList;
import java.util.ArrayList;
import java.util.Queue;
import java.util.List;

class TreeNode {
  int value;
  TreeNode left;
  TreeNode right;

  TreeNode(int value) {
    this.value = value;
  }
}

class ReverseLevelOrderTraversal {
  public static List<List<Integer>> reverseTraversal(TreeNode root) {
    List<List<Integer>> resultReverseNodes = new LinkedList<List<Integer>>();
    if (root == null)
      return resultReverseNodes;
    Queue<TreeNode> queue = new LinkedList<>();
    queue.offer(root);
    while (!queue.isEmpty()) {
      List<Integer> currentList = new ArrayList<>();
      int levelSize = queue.size();
      for (int i = 0; i < levelSize; i++) {
        TreeNode currentNode = queue.poll();
        currentList.add(currentNode.value);
        if (currentNode.left != null)
          queue.offer(currentNode.left);
        if (currentNode.right != null)
          queue.offer(currentNode.right);
      }
      resultReverseNodes.add(0, currentList);
    }
    return resultReverseNodes;
  }

  public static void main(String[] args) {
    TreeNode root = new TreeNode(1);
    root.left = new TreeNode(2);
    root.right = new TreeNode(3);
    root.left.right = new TreeNode(4);
    root.left.left = new TreeNode(5);
    root.right.left = new TreeNode(6);
    List<List<Integer>> reverseNodes = ReverseLevelOrderTraversal.reverseTraversal(root);
    System.out.println(reverseNodes); // [[5,4,6],[2,3],[1]];
  }
}
